### PR TITLE
The I/O Feature Preview

### PR DESCRIPTION
**Pull request by mikbras**
This PR provides a preview of the Open Enclave I/O feature. This feature adds:
- Socket I/O
- File I/O (secure and non-secure)

The PR was so big (due 3rdparty/linux-sgx) that we created a version that omits that directory.

### CODE REVIEW

**Comment by johnkord** [comment]
Added the "do not merge" label to this because this branch is solely for reviewing. If you want to build and test this PR, use the "feature.io" branch, I think :)

**Comment by johnkord** [comment]
In case this isn't an already known issue, I hit two test errors with the "fs2" and "resolver" tests when building and testing on Ubuntu 16.04:

fs2 failure:

```
   Start 97: tests/fs2

97: Test command: /home/johnkord/openenclave/build/tests/fs/host/fs_host "/home/johnkord/openenclave/build/tests/fs/enc/fs_enc" "/home/johnkord/openenclave" "/home/johnkord/openenclave/build/tests/fs/tmp"
97: Test timeout computed to be: 10000000
97: Test failed: ../tests/fs/enc/enc.cpp(481): test_fs oe_mkdir_d(OE_DEVID_HOSTFS, tmp_dir, 0777) == 0
97: Test failed: ../tests/fs/host/host.c(33): main r == OE_OK
97: 18:49:01:464011 tid(0x7fb131ede740) (H)[ERROR]:OE_ENCLAVE_ABORTING[../host/sgx/calls.c oe_call_enclave_function:779]
1/1 Test #97: tests/fs2 ........................Child aborted***Exception:   0.31 sec

0% tests passed, 1 tests failed out of 1

Total Test time (real) =   0.32 sec
```


resolver test:

```
    Start 167: tests/resolver

167: Test command: /home/johnkord/openenclave/build/tests/resolver/host/resolver_host "/home/johnkord/openenclave/build/tests/resolver/enc/resolver_enc"
167: Test timeout computed to be: 10000000
*** Error in `/home/johnkord/openenclave/build/tests/resolver/host/resolver_host': corrupted size vs. prev_size: 0x0000000000e0a1b0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7f7dfcfa77e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x7e9dc)[0x7f7dfcfae9dc]
/lib/x86_64-linux-gnu/libc.so.6(+0x81cde)[0x7f7dfcfb1cde]
/lib/x86_64-linux-gnu/libc.so.6(__libc_malloc+0x54)[0x7f7dfcfb4184]
/home/johnkord/openenclave/build/tests/resolver/host/resolver_host(HandleMalloc+0x67)[0x41bf37]
/home/johnkord/openenclave/build/tests/resolver/host/resolver_host[0x40987a]
/home/johnkord/openenclave/build/tests/resolver/host/resolver_host(__oe_dispatch_ocall+0x281)[0x409411]
/home/johnkord/openenclave/build/tests/resolver/host/resolver_host[0x4285cc]
/home/johnkord/openenclave/build/tests/resolver/host/resolver_host(__morestack+0xa9)[0x428514]
======= Memory map: ========
00400000-0044f000 r-xp 00000000 08:01 537768                             /home/johnkord/openenclave/build/tests/resolver/host/resolver_host
0064e000-0064f000 r-xp 0004e000 08:01 537768                             /home/johnkord/openenclave/build/tests/resolver/host/resolver_host
0064f000-00650000 rwxp 0004f000 08:01 537768                             /home/johnkord/openenclave/build/tests/resolver/host/resolver_host
00650000-00652000 rwxp 00000000 00:00 0
00de1000-00e36000 rwxp 00000000 00:00 0                                  [heap]
7f7df4000000-7f7df4021000 rwxp 00000000 00:00 0
7f7df4021000-7f7df8000000 ---p 00000000 00:00 0
7f7dfb000000-7f7dfb15d000 r-xs 00000000 00:06 150                        /dev/sgx
7f7dfb15d000-7f7dfb35c000 ---s 0015d000 00:06 150                        /dev/sgx
7f7dfb35c000-7f7dfb368000 rwxs 0035c000 00:06 150                        /dev/sgx
7f7dfb368000-7f7dfb36c000 r-xs 00368000 00:06 150                        /dev/sgx
7f7dfb36c000-7f7dfb46c000 rwxs 0036c000 00:06 150                        /dev/sgx
7f7dfb46c000-7f7dfb46d000 ---s 0046c000 00:06 150                        /dev/sgx
7f7dfb46d000-7f7dfb56d000 rwxs 0046d000 00:06 150                        /dev/sgx
7f7dfb56d000-7f7dfb56e000 ---s 0056d000 00:06 150                        /dev/sgx
7f7dfb56e000-7f7dfb571000 rwxs 0056e000 00:06 150                        /dev/sgx
7f7dfb571000-7f7dfb572000 ---s 00571000 00:06 150                        /dev/sgx
7f7dfb572000-7f7dfb574000 rwxs 00572000 00:06 150                        /dev/sgx
7f7dfb574000-7f7dfb800000 ---s 00574000 00:06 150                        /dev/sgx
7f7dfbf18000-7f7dfbf31000 r-xp 00000000 08:01 2150                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f7dfbf31000-7f7dfc130000 ---p 00019000 08:01 2150                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f7dfc130000-7f7dfc131000 r-xp 00018000 08:01 2150                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f7dfc131000-7f7dfc132000 rwxp 00019000 08:01 2150                       /lib/x86_64-linux-gnu/libz.so.1.2.8
7f7dfc132000-7f7dfc23a000 r-xp 00000000 08:01 27381                      /lib/x86_64-linux-gnu/libm-2.23.so
7f7dfc23a000-7f7dfc439000 ---p 00108000 08:01 27381                      /lib/x86_64-linux-gnu/libm-2.23.so
7f7dfc439000-7f7dfc43a000 r-xp 00107000 08:01 27381                      /lib/x86_64-linux-gnu/libm-2.23.so
7f7dfc43a000-7f7dfc43b000 rwxp 00108000 08:01 27381                      /lib/x86_64-linux-gnu/libm-2.23.so
7f7dfc43b000-7f7dfc555000 r-xp 00000000 08:01 35118                      /usr/lib/x86_64-linux-gnu/libprotobuf.so.9.0.1
7f7dfc555000-7f7dfc754000 ---p 0011a000 08:01 35118                      /usr/lib/x86_64-linux-gnu/libprotobuf.so.9.0.1
7f7dfc754000-7f7dfc758000 r-xp 00119000 08:01 35118                      /usr/lib/x86_64-linux-gnu/libprotobuf.so.9.0.1
7f7dfc758000-7f7dfc759000 rwxp 0011d000 08:01 35118                      /usr/lib/x86_64-linux-gnu/libprotobuf.so.9.0.1
7f7dfc759000-7f7dfc76f000 r-xp 00000000 08:01 1989                       /lib/x86_64-linux-gnu/libgcc_s.so.1
7f7dfc76f000-7f7dfc96e000 ---p 00016000 08:01 1989                       /lib/x86_64-linux-gnu/libgcc_s.so.1
7f7dfc96e000-7f7dfc96f000 rwxp 00015000 08:01 1989                       /lib/x86_64-linux-gnu/libgcc_s.so.1
7f7dfc96f000-7f7dfcae1000 r-xp 00000000 08:01 25949                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7f7dfcae1000-7f7dfcce1000 ---p 00172000 08:01 25949                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7f7dfcce1000-7f7dfcceb000 r-xp 00172000 08:01 25949                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7f7dfcceb000-7f7dfcced000 rwxp 0017c000 08:01 25949                      /usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.21
7f7dfcced000-7f7dfccf1000 rwxp 00000000 00:00 0
7f7dfccf1000-7f7dfcd2d000 r-xp 00000000 08:01 35136                      /usr/lib/x86_64-linux-gnu/libsgx_uae_service.so.1.0.102.0
7f7dfcd2d000-7f7dfcf2c000 ---p 0003c000 08:01 35136                      /usr/lib/x86_64-linux-gnu/libsgx_uae_service.so.1.0.102.0
7f7dfcf2c000-7f7dfcf2f000 r-xp 0003b000 08:01 35136                      /usr/lib/x86_64-linux-gnu/libsgx_uae_service.so.1.0.102.0
7f7dfcf2f000-7f7dfcf30000 rwxp 0003e000 08:01 35136                      /usr/lib/x86_64-linux-gnu/libsgx_uae_service.so.1.0.102.0
7f7dfcf30000-7f7dfd0f0000 r-xp 00000000 08:01 27378                      /lib/x86_64-linux-gnu/libc-2.23.so
7f7dfd0f0000-7f7dfd2f0000 ---p 001c0000 08:01 27378                      /lib/x86_64-linux-gnu/libc-2.23.so
7f7dfd2f0000-7f7dfd2f4000 r-xp 001c0000 08:01 27378                      /lib/x86_64-linux-gnu/libc-2.23.so
7f7dfd2f4000-7f7dfd2f6000 rwxp 001c4000 08:01 27378                      /lib/x86_64-linux-gnu/libc-2.23.so
7f7dfd2f6000-7f7dfd2fa000 rwxp 00000000 00:00 0
7f7dfd2fa000-7f7dfd318000 r-xp 00000000 08:01 35132                      /usr/lib/x86_64-linux-gnu/libsgx_urts.so.1.0.102.0
7f7dfd318000-7f7dfd518000 ---p 0001e000 08:01 35132                      /usr/lib/x86_64-linux-gnu/libsgx_urts.so.1.0.102.0
7f7dfd518000-7f7dfd519000 r-xp 0001e000 08:01 35132                      /usr/lib/x86_64-linux-gnu/libsgx_urts.so.1.0.102.0
7f7dfd519000-7f7dfd51b000 rwxp 0001f000 08:01 35132                      /usr/lib/x86_64-linux-gnu/libsgx_urts.so.1.0.102.0
7f7dfd51b000-7f7dfd51f000 rwxp 00000000 00:00 0
7f7dfd51f000-7f7dfd529000 r-xp 00000000 08:01 35111                      /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1.0.101.48192
7f7dfd529000-7f7dfd728000 ---p 0000a000 08:01 35111                      /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1.0.101.48192
7f7dfd728000-7f7dfd729000 r-xp 00009000 08:01 35111                      /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1.0.101.48192
7f7dfd729000-7f7dfd72a000 rwxp 0000a000 08:01 35111                      /usr/lib/x86_64-linux-gnu/libsgx_dcap_ql.so.1.0.101.48192
7f7dfd72a000-7f7dfd72e000 r-xp 00000000 08:01 35135                      /usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1.0.102.0
7f7dfd72e000-7f7dfd92d000 ---p 00004000 08:01 35135                      /usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1.0.102.0
7f7dfd92d000-7f7dfd92e000 r-xp 00003000 08:01 35135                      /usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1.0.102.0
7f7dfd92e000-7f7dfd92f000 rwxp 00004000 08:01 35135                      /usr/lib/x86_64-linux-gnu/libsgx_enclave_common.so.1.0.102.0
7f7dfd92f000-7f7dfd947000 r-xp 00000000 08:01 27346                      /lib/x86_64-linux-gnu/libpthread-2.23.so
7f7dfd947000-7f7dfdb46000 ---p 00018000 08:01 27346                      /lib/x86_64-linux-gnu/libpthread-2.23.so
7f7dfdb46000-7f7dfdb47000 r-xp 00017000 08:01 27346                      /lib/x86_64-linux-gnu/libpthread-2.23.so
7f7dfdb47000-7f7dfdb48000 rwxp 00018000 08:01 27346                      /lib/x86_64-linux-gnu/libpthread-2.23.so
7f7dfdb48000-7f7dfdb4c000 rwxp 00000000 00:00 0
7f7dfdb4c000-7f7dfdb4f000 r-xp 00000000 08:01 27344                      /lib/x86_64-linux-gnu/libdl-2.23.so
7f7dfdb4f000-7f7dfdd4e000 ---p 00003000 08:01 27344                      /lib/x86_64-linux-gnu/libdl-2.23.so
7f7dfdd4e000-7f7dfdd4f000 r-xp 00002000 08:01 27344                      /lib/x86_64-linux-gnu/libdl-2.23.so
7f7dfdd4f000-7f7dfdd50000 rwxp 00003000 08:01 27344                      /lib/x86_64-linux-gnu/libdl-2.23.so
7f7dfdd50000-7f7dfdf6b000 r-xp 00000000 08:01 3898                       /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
7f7dfdf6b000-7f7dfe16a000 ---p 0021b000 08:01 3898                       /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
7f7dfe16a000-7f7dfe186000 r-xp 0021a000 08:01 3898                       /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
7f7dfe186000-7f7dfe192000 rwxp 00236000 08:01 3898                       /lib/x86_64-linux-gnu/libcrypto.so.1.0.0
7f7dfe192000-7f7dfe195000 rwxp 00000000 00:00 0
7f7dfe195000-7f7dfe1bb000 r-xp 00000000 08:01 27345                      /lib/x86_64-linux-gnu/ld-2.23.so
7f7dfe3a6000-7f7dfe3b0000 rwxp 00000000 00:00 0
7f7dfe3b9000-7f7dfe3ba000 rwxp 00000000 00:00 0
7f7dfe3ba000-7f7dfe3bb000 r-xp 00025000 08:01 27345                      /lib/x86_64-linux-gnu/ld-2.23.so
7f7dfe3bb000-7f7dfe3bc000 rwxp 00026000 08:01 27345                      /lib/x86_64-linux-gnu/ld-2.23.so
7f7dfe3bc000-7f7dfe3bd000 rwxp 00000000 00:00 0
7ffd3f909000-7ffd3f92a000 rwxp 00000000 00:00 0                          [stack]
7ffd3f947000-7ffd3f94a000 r--p 00000000 00:00 0                          [vvar]
7ffd3f94a000-7ffd3f94c000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
167: bufptr(end)-bufptr(start) = 192
1/1 Test #167: tests/resolver ...................Child aborted***Exception:   0.19 sec

0% tests passed, 1 tests failed out of 1
```

File: enclave/core/cwd.c [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+#include <openenclave/corelibc/errno.h>
+#include <openenclave/corelibc/limits.h>
+#include <openenclave/corelibc/pthread.h>
+#include <openenclave/corelibc/stdio.h>
+#include <openenclave/corelibc/stdlib.h>
+#include <openenclave/corelibc/string.h>
+#include <openenclave/corelibc/sys/stat.h>
+
+static char _cwd[OE_PATH_MAX] = "/";
```
Comment (johnkord) [inline enclave/core/cwd.c]: Do you think by default this should be initialized to the cwd of the host?

Comment (johnkord) [reply enclave/core/cwd.c]: Oh I see, this is abstracted out by the device that gets mounted to a specific path, forget my comment above. Interesting!

**Comment by soccerGB** [comment]
I agree with John that:  If someone wants to build and test this PR on their dev system, they should use the "feature.io" branch because "feature.io.no-linux-sgx does not build"

**Comment by andschwa** [comment]
Closing per triage meeting as this has been split into reviewable PRs. Thanks!

**Pull request closed without merging**